### PR TITLE
Add --override-module-licence and --check-go-modules options

### PR DESCRIPTION
**Pull request by kragniz**
This allows you to override the licence for a given go module. This
finds the directory for the module, then continues with compliance
checking normally.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -46,6 +50,19 @@ func init() {
 func validateCompliance(_ *cobra.Command, args []string) {
 	setLogLevel(logLevel)
 
+	for module, licence := range overriddenModuleLicences {
+		cmd := exec.Command("go", "list", "-m", "-f", "\"{{.Dir}}\"", module)
+		cmd.Env = os.Environ()
+		cmd.Env = append(cmd.Env, "GO111MODULE=on")
+
+		out, err := cmd.CombinedOutput()
```
Comment (wallrj) [inline main.go]: Think this should be `.Output()` just in case `go list` prints unrelated stuff to stderr.

File: main.go [hunk 2]
```diff
@@ -46,6 +50,19 @@ func init() {
 func validateCompliance(_ *cobra.Command, args []string) {
 	setLogLevel(logLevel)
 
+	for module, licence := range overriddenModuleLicences {
+		cmd := exec.Command("go", "list", "-m", "-f", "\"{{.Dir}}\"", module)
+		cmd.Env = os.Environ()
+		cmd.Env = append(cmd.Env, "GO111MODULE=on")
```
Comment (wallrj) [inline main.go]: Is this necessary? Presumably this must be set anyway, by anyone calling `licence-compliance-checker --override-module-licence`

File: README.md [hunk 3]
```diff
@@ -35,7 +35,8 @@ Input argument | Meaning
 ---------|---------
 --restricted-licence (-r) | The name of the licence to restrict. Repeat this flag to specify multiple values.
 --ignore-project (-i) | Project which licence will not be checked for compliance. Repeat this flag to specify multiple values.
---override-licence (-o) | Can be used to override the licence detected for a project - e.g. github.com/spf13/cobra=MIT. Repeat this flag to specify multiple values.
+--override-licence (-o) | Can be used to override the licence detected for a project directory - e.g. vendor/github.com/spf13/cobra=MIT. Repeat this flag to specify multiple values.
```
Comment (sebbonnet) [inline README.md]: These flags are now mutually exclusive: `override-licence`, `override-licence-module`.
We should mention it and also also validation to reject invokation when both flags are used

Comment (sebbonnet) [reply README.md]: Can we have usage examples for both flags in the README? e.g. when using vendor, when using go modules

File: main.go [hunk 4]
```diff
@@ -35,23 +38,56 @@ func main() {
...
 	}
 
+	if checkGoModules {
+		if len(args) > 0 {
+			logAndExit("--check-go-modules and positional args cannot be set at the same time (received %d)", len(args))
+		}
+
+		var err error
+		args, err = getGoModules()
+		if err != nil {
```
Comment (sebbonnet) [inline main.go]: I would add a log.Info here to print the modules found. 
This would then be displayed when using `-L INFO`

File: test/e2e/e2e_test.go [hunk 5]
```diff
@@ -127,6 +143,62 @@ var _ = Describe("License Compliance Checker", func() {
 		})
 	})
 
+	Context("modules", func() {
+		It("should check a project's modules", func() {
```
Comment (sebbonnet) [inline test/e2e/e2e_test.go]: I would add a test showing how it fails when `GO111MODULE` is not provided

Comment (sebbonnet) [reply test/e2e/e2e_test.go]: Can we also add a test showing a usage without the `check-go-modules` option?

File: test/e2e/testdata/go-module/go.mod [hunk 6]
```diff
@@ -0,0 +1,5 @@
+module github.com/sky-uk/licence-compliance-checker/e2e/testdata/go-module
+
+go 1.12
```
Comment (sebbonnet) [inline test/e2e/testdata/go-module/go.mod]: Is that needed?

File: Makefile [hunk 7]
```diff
@@ -12,6 +12,7 @@ setup :
 	go get -v golang.org/x/lint/golint golang.org/x/tools/cmd/goimports github.com/golang/dep/cmd/dep
 	go get -v github.com/onsi/ginkgo/ginkgo && cd $$GOPATH/src/github.com/onsi/ginkgo && git checkout 'v1.6.0' && go install github.com/onsi/ginkgo/ginkgo
 	dep ensure -v -vendor-only
+	cd test/e2e/testdata/go-module && GO111MODULE=on go build
```
Comment (sebbonnet) [inline Makefile]: Is this needed?

File: Makefile [hunk 8]
```diff
@@ -28,7 +29,7 @@ licencecheck:
 	set -e ;\
  	restricted=$$(paste -s -d ',' restricted-licences.txt) ;\
  	projects=$$(dep status -f='vendor/{{ .ProjectRoot }} ') ;\
- 	$(BUILD_DIR)/bin/licence-compliance-checker -L error -A -r $$restricted $$projects ;
+ 	$(BUILD_DIR)/bin/licence-compliance-checker -L error -A -r $$restricted -m github.com/spf13/cobra=MIT $$projects ;
```
Comment (sebbonnet) [inline Makefile]: I don't think this is needed. 
I can run this target fine without the extra `-m github.com/spf13/cobra=MIT `
Travis also complains about it, but it actually fails to list the modules probably because GO111MODULE isn't enabled
```
ERRO[0000] Failed to list go module github.com/spf13/cobra: output: , error: exit status 1 
```

File: main.go [hunk 9]
```diff
@@ -35,23 +38,56 @@ func main() {
...
+	rootCmd.PersistentFlags().StringToStringVarP(&overriddenModuleLicences, "override-module-licence", "m", map[string]string{}, "can be used to override the licence detected for a go module - e.g. github.com/spf13/cobra=MIT. Repeat this flag to specify multiple values.")
 	rootCmd.PersistentFlags().StringSliceVarP(&restrictedLicences, "restricted-licence", "r", []string{}, "licence that will fail the compliance check if found for a project. Repeat this flag to specify multiple values.")
 	rootCmd.PersistentFlags().StringVarP(&logLevel, "log-level", "L", "", "(output) should be one of: (none), debug, info, warn, error, fatal, panic. default (none)")
 	rootCmd.PersistentFlags().BoolVarP(&showComplianceAll, "show-compliance-all", "A", false, "(output) to show compliance checks as JSON regardless of outcome")
 	rootCmd.PersistentFlags().BoolVarP(&showComplianceErrors, "show-compliance-errors", "E", false, "(output) to show compliance checks as JSON only in case of errors")
+	rootCmd.PersistentFlags().BoolVarP(&checkGoModules, "check-go-modules", "", false, "check all go modules a project depends on")
```
[the remark this answers is not in the document]
Comment (sebbonnet) [reply main.go]: The same description as in the README would be fine

File: main.go [hunk 10]
```diff
@@ -35,23 +38,56 @@ func main() {
...
+	if len(overriddenModuleLicences) > 0 && len(overriddenLicences) > 0 {
+		logAndExit("Only use one of --override-module-licence (%d uses) and --override-licence (%d uses)", len(overriddenModuleLicences), len(overriddenLicences))
+	}
+
+	for module, licence := range overriddenModuleLicences {
+		cmd := exec.Command("go", "list", "-m", "-f", "\"{{.Dir}}\"", module)
+
+		out, err := cmd.Output()
+		if err != nil {
+			logAndExit("Failed to list go module %s: output: %s, error: %s", module, out, err)
```
[the remark this answers is not in the document]
Comment (kragniz) [reply main.go]: Good idea

File: test/e2e/e2e_test.go (hunk 5 above)
Comment (kragniz) [reply test/e2e/e2e_test.go]: testing when `GO111MODULE` is not set is a little more awkward, since that behaviour depends on whether the tests are being run from inside the GOPATH

**Comment by kragniz** [comment]
@sebbonnet I think I've fixed all those up, mind taking another look?

**Comment by sebbonnet** [comment]
Thanks @kragniz - I've released this under v1.0.1
